### PR TITLE
perf(ethereum): provide redis option for token client

### PR DESCRIPTION
**Pull request by kallydev**
## Summary

Related to #202.

## Checklist

- [x] The commit message follows [Angular Contributing guidelines](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit);
- [ ] Tests for the changes have been added (for bug fixes / features);

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by polebug — APPROVED** [review]

**Comment by kallydev** [comment]
TODO list @pseudoyu.

- [ ] Load the Redis configuration and initialize the Redis client.
- [ ] Use the `WithRueidisClient(rueidisClient rueidis.Client) Option` option for Ethereum's fallback worker.

File: provider/ethereum/token/client.go [hunk 1]
```diff
@@ -279,9 +302,34 @@ func (c *client) lookupNFT(ctx context.Context, chain uint64, address common.Add
...
+	// Cache the token metadata to Redis.
+	if c.rueidisClient != nil {
+		if value, err := json.Marshal(tokenMetadata); err == nil {
+			command := c.rueidisClient.B().Setex().
+				Key(c.buildCacheKey(chainID, address, id)).
+				Seconds(int64(defaultCacheDuration.Seconds())).
+				Value(string(value)).
+				Build()
+
+			c.rueidisClient.Do(ctx, command)
```
Comment (HenryQW) [inline provider/ethereum/token/client.go:314]: missing error check

Comment (HenryQW) [reply provider/ethereum/token/client.go:314]: repeated logic: see L179-190.
it's better to have an abstract func cacheTokenMetadata

File: provider/ethereum/token/client.go [hunk 2]
```diff
@@ -27,17 +28,18 @@ import (
 	"github.com/tdewolff/minify/v2/minify"
 	"github.com/tidwall/gjson"
 	"github.com/vincent-petithory/dataurl"
-	"go.opentelemetry.io/otel"
-	"go.opentelemetry.io/otel/attribute"
-	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	defaultCacheDuration = 24 * time.Hour
```
Comment (HenryQW) [inline provider/ethereum/token/client.go:34]: I am okay with this not being loaded from env, but is 24H too short? I don't think the majority ever updates metadata.

File: provider/ethereum/token/client.go [hunk 3]
```diff
@@ -243,10 +248,28 @@ func (c *client) lookupERC20ByRPC(ctx context.Context, chainID uint64, address c
 	return &tokenMetadata, nil
 }
 
+func (c *client) lookupERC20ByRedis(ctx context.Context, chainID uint64, address common.Address) (*metadata.Token, error) {
```
Comment (HenryQW) [inline provider/ethereum/token/client.go:251]: all func names are very similar, so adding a line of comment for all will be useful.

File: provider/ethereum/token/client.go [hunk 4]
```diff
@@ -587,12 +635,32 @@ func (c *client) standardizeNonFungibleTokenMetadata(_ context.Context, nonFungi
...
 // NewClient returns a new client.
-func NewClient(ethereumClient ethereum.Client) Client {
+func NewClient(ethereumClient ethereum.Client, options ...Option) Client {
 	instance := client{
 		ethereumClient: ethereumClient,
 	}
 
+	for _, option := range options {
+		if err := option(&instance); err != nil {
+			panic(err)
```
Comment (HenryQW) [inline provider/ethereum/token/client.go:660]: should it panic here?

File: provider/ethereum/token/client.go [hunk 5]
```diff
@@ -587,12 +635,32 @@ func (c *client) standardizeNonFungibleTokenMetadata(_ context.Context, nonFungi
...
+	}
+
+	return key
+}
+
 // NewClient returns a new client.
-func NewClient(ethereumClient ethereum.Client) Client {
+func NewClient(ethereumClient ethereum.Client, options ...Option) Client {
 	instance := client{
 		ethereumClient: ethereumClient,
```
Comment (HenryQW) [inline provider/ethereum/token/client.go:655]: We need to use fx for dep injection here.

Comment (HenryQW) [reply provider/ethereum/token/client.go:655]: probably a separate PR.

File: deploy/config.yaml [hunk 6]
```diff
@@ -19,6 +19,14 @@ stream:
   topic: rss3.node.feeds
   uri: localhost:9092
 
+redis:
```
[the remark this answers is not in the document]
Comment (pseudoyu) [reply deploy/config.yaml:22]: tls config added.

File: provider/ethereum/token/client.go (hunk 2 above)
Comment (pseudoyu) [reply provider/ethereum/token/client.go]: Changed to 7 days.

File: provider/ethereum/token/client.go (hunk 5 above)
Comment (pseudoyu) [reply provider/ethereum/token/client.go:665]: Yes will be a separate fx support pr.

**Review by HenryQW — APPROVED** [review]

**Pull request merged**